File: sendTx.ts
import { config } from "dotenv";
import { ethers } from "ethers";

config();

async function main() {
    const rpcUrl = process.env.RPC_URL!;///第三方節點
    const privateKey = process.env.PRIVATE_KEY!;///測試錢包私鑰

    const provider = new ethers.JsonRpcProvider(rpcUrl);///連接到Sepolia節點
    const wallet = new ethers.Wallet(privateKey, provider);//使用私鑰創建錢包

    console.log("0xbd166ac8bCD24cdB575440817C995Ce113717a21:", wallet.address);//錢包地址

    const to = wallet.address;//發送到自己的地址

    const tx = await wallet.sendTransaction({//發送交易
        to,
        value: ethers.parseEther("0.01"),//發送0.01 ETH
    });

    console.log("Transaction hash:", tx.hash);//交易hash

    const receipt = await tx.wait();

    console.log("交易確認！");
    console.log("區塊號:", receipt?.blockNumber);
    console.log("Gas Used:", receipt?.gasUsed.toString());//使用的Gas量手續費
    }

    main().catch(console.error);

File: src/bin/send_tx.rs
use dotenv::dotenv;
use ethers::{
    providers::{Http, Middleware, Provider},
    signers::{LocalWallet, Signer},
    types::{TransactionRequest, U256},
    utils::parse_ether,
};
use std::{env, str::FromStr, sync::Arc};

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    dotenv().ok();

    let rpc_url = env::var("RPC_URL").expect("RPC_URL is required");
    let private_key = env::var("PRIVATE_KEY").expect("PRIVATE_KEY is required");

    let provider = Provider::<Http>::try_from(rpc_url)?;
    let wallet = LocalWallet::from_str(&private_key)?;
    let wallet = wallet.with_chain_id(11155111u64);
    let address = wallet.address();
    let client = Arc::new(ethers::middleware::SignerMiddleware::new(provider, wallet));

    println!("Wallet address: {address:?}");

    let tx = TransactionRequest::new()
        .to(address)
        .value(parse_ether("0.01")?)
        .from(address);

    let pending = client.send_transaction(tx, None).await?;
    println!("Transaction hash: {:?}", pending.tx_hash());

    let receipt = pending.await?;
    if let Some(receipt) = receipt {
        println!("Transaction confirmed");
        println!("Block number: {:?}", receipt.block_number);
        println!("Gas used: {:?}", receipt.gas_used.unwrap_or(U256::from(0)));
    }

    Ok(())
}
